### ISSUE TITLE
[QUESTION] How to evaluate training loss before the first epoch?

### ISSUE THREAD

**Issue by IvoFelix** [unit8co/darts#2704]
I wish to stop training when a certain threshold of training loss is reached.

Initially I tried to achieve this with Pytorch Lightning's EarlyStopping callback:
```
EarlyStopping(monitor="train_loss", stopping_threshold=0.01)
```

This works, however when I am finetuning the model and I load the weights from the previous model I wish the training to stop before the first epoch in case the initial weights are already good enough for the training loss to be below the threshold. This doesn't seem possible using `EarlyStopping`, as it uses `on_train_epoch_end()` and `on_validation_end()` so it only runs after the first epoch.

So I then attempted to create my own callback:
```
class TargetMetricStopping(Callback):

    def __init__(self, target_metric: str, target_loss: float):
        self.target_metric = target_metric
        self.target_loss = target_loss

    def on_train_epoch_start(self, trainer, pl_module):
        current_loss = trainer.callback_metrics.get(self.target_metric, float('inf'))
        if current_loss <= self.target_loss:
            print(f"Stopping early due to target {self.target_metric} of {self.target_loss} reached.")
            trainer.should_stop = True  # stop training
```

However, this does not help either as `trainer.callback_metrics` is empty before the end of the first epoch.

How can I get the training loss before the first epoch?

**Comment by IvoFelix** [comment]
Thanks for the reply.

Even if the training loss were saved with the checkpoint it would be the training loss for the dataset used to train the initial model, not for the finetuning dataset which contains extra data, isn't that so?

About triggering a round of prediction and loss computation, I tried it as well but I'm not familiar with Pytorch Lightning and was unsucessful so far. I added an `on_train_start()` method, but I'm not sure on what input to pass to pl_module:
```
    def on_train_start(self, trainer, pl_module):

        pl_module.eval()
        total_loss = 0.0
        with torch.no_grad():
            for batch in trainer.train_dataloader:
                x = batch[0]
                y = batch[1]
                y_hat = pl_module(x)
                loss = nn.functional.mse_loss(y_hat, y)
                total_loss += loss.item()
        avg_loss = total_loss / len(trainer.train_dataloader)
        pl_module.train()

        # Skip training if the loss is below the threshold
        if avg_loss < self.threshold:
            trainer.should_stop = True
```

**Comment by madtoinou** [comment]
> Even if the training loss were saved with the checkpoint it would be the training loss for the dataset used to train the initial model, not for the finetuning dataset which contains extra data, isn't that so?

Correct.

> About triggering a round of prediction and loss computation, I tried it as well but I'm not familiar with Pytorch Lightning and was unsucessful so far. I added an `on_train_start()` method, but I'm not sure on what input to pass to pl_module:

Honestly, the simplest approach would be to load the checkpoint and have a helper method to compute the loss outside of the training instead of relying on callbacks, and proceed with the fine-tuning only if the loss is above the threshold.

Alternatively, as mentioned in my previous message, you could reproduce some of the `EarlyStopping` logic to occur on `on_train_batch_start` instead of the native `on_train_epoch_end()`/`on_validation_end()`.

**Comment by IvoFelix** [comment]
> Honestly, the simplest approach would be to load the checkpoint and have a helper method to compute the loss outside of the training instead of relying on callbacks, and proceed with the fine-tuning only if the loss is above the threshold.

I see, what would be the simplest approach for this?

I've tried with `historical_forecasts()`:
```
predictions = forecast_model.historical_forecasts(
    series=target_series_list,
    past_covariates=covariate_series_list,
    forecast_horizon=1,
    retrain=False,
    verbose=True,
)

training_loss = mse(target_series_list, predictions, series_reduction=np.nanmean)
```
But the result is inconsistent with the training loss I obtained from the initial training and I'm not sure what am I missing. Is there a better way?

**Comment by madtoinou** [comment]
The "training loss" and "historical forecasts" are two very different things; the later corresponds to "how would the model have performed if it was deployed during this period, as the data is being recorded".

In order to generate a value comparable with the training loss, you need to mimic a batch of prediction on the data. One way to achieve it, would be something like:

```python
import numpy as np

from darts.datasets import AirPassengersDataset
from darts.models import NLinearModel
from darts import TimeSeries
from darts.metrics import mse

# fit the model on the initial dataset
series = AirPassengersDataset().load().astype("float32")
model = NLinearModel(input_chunk_length=3, output_chunk_length=1)
model.fit(series, epochs=2)
model.save("test.pt")

# load the weights (recommended for the fine-tuning)
model_ft = NLinearModel(input_chunk_length=3, output_chunk_length=1)
model_ft.load_weights(path="test.pt")

# mimic the train dataset used in `fit()`
dataset = model_ft._build_train_dataset(series, None, None, None, max_samples_per_ts=16)

# perform a batch of prediction
preds = []
truths = []
for entry in dataset:
    past_target, _, _, _, _, _, future_target = entry
    input_series = TimeSeries.from_values(past_target)
    pred = model.predict(n=1, series=input_series, verbose=False)
    preds.append(pred)
    truths.append(TimeSeries.from_times_and_values(pred.time_index, future_target))
print(mse(truths, preds, series_reduction=np.mean))
```

**Comment by madtoinou** [comment]
Are you using exactly the same series?  The same loss function? Depending on the number of samples, the value can be quite different as well.

To investigate the stability of you model, I would recommend during historical forecast on the training series to make sure that your accuracy/performance is uniform over time. If it is not, it would explain the different between the two.

**Comment by IvoFelix** [comment]
I meant while using the very example you posted, I get different training losses for the initial training and for the final value. Can you confirm if you got the same training losses?

**Comment by madtoinou** [comment]
By replacing the `NLinearModel` with the `DLinearModel`, and considerably increasing the number of epochs, the training loss and "synthetic" training loss are very close:

```python
model = DLinearModel(input_chunk_length=12, output_chunk_length=1)
model.fit(series, epochs=200)
>>>  378.0
# [...]
# increase the number of sample to 64 to make it more stable
dataset = model_ft._build_train_dataset(series, None, None, None, max_samples_per_ts=64)
# [...]
mse(truths, preds, series_reduction=np.mean))
>>> 304.8745
```

**Comment by IvoFelix** [comment]
Thank you so for your assistance so far.

Shouldn't it be possible to get the same training loss while using the same input data, network architecture, weights and model parameters? What am I missing?

**Comment by madtoinou** [comment]
Not entirely sure to understand your questions but since the training loss is computed on a batch, unless you get exactly the same batch, the values obtained will never be exactly the same (and is theoretically as representative as the last value observed during the training).

Considering your original objective, I think that the solution presented above is the best workaround (the cleanest being trigerring the fine-tuning anyway and eventually stopping it using the `EarlyStopper` if relevant) and if you are fine with it, I would like to close the issue.

**Comment by IvoFelix** [comment]
Sorry for taking some time to reply, I was really pressed for time last week.

> Not entirely sure to understand your questions but since the training loss is computed on a batch, unless you get exactly the same batch, the values obtained will never be exactly the same (and is theoretically as representative as the last value observed during the training).

Of course the loss is computed for each batch, and then a global training loss is also evaluated based on the loss values for each batch, which is what is used for the `EarlyStopping` callback, so of course I was talking about the global training loss, which is what is displayed in the progress bar at the end of the epoch.

In addition I can get the same exact global training loss every single time I train with the same input, as long as I fix the seed for any RNG used by the model using:
```
import numpy as np
import torch
import random

torch.manual_seed(42)
np.random.seed(42)
random.seed(42)
```
This makes the results reproducible so that I can get the same exact results every single time, so I see no reason why the same could not be done for computing the global training loss that was obtained previously before training again for finetuning.

> Considering your original objective, I think that the solution presented above is the best workaround (the cleanest being trigerring the fine-tuning anyway and eventually stopping it using the EarlyStopper if relevant) and if you are fine with it, I would like to close the issue.

That is the obvious choice which is why it was the first thing I attempted. Unfortunately, it does not solve my problem. If I use the `EarlyStopping` callback the training loss is only evaluated after the first epoch which, even for the exact same data, may unnecessarily result in a different training loss that is above the initial threshold and the model may now take numerous epochs to converge again to a training loss below the intended threshold. I am training a significant number of models, and these extra epochs add up to a significant waste of processing time.

As for your solution, since the training loss also does not match the initial training loss, it is subject to the same risk.
